### ISSUE TITLE
CustomDialog ManagedUI fails to go to next screen

### ISSUE THREAD

**Issue by kriactivedesigns** [oleg-shilo/wixsharp#731]
```
project.ManagedUI.InstallDialogs.Add<CustomInstallDialogs.WelcomeDialog>()
                                                     .Add<CustomInstallDialogs.LicenceDialog>()
                                                     .Add<CustomInstallDialogs.FeaturesDialog>()
                                                     .Add<CustomInstallDialogs.ProgressDialog>()
                                                     .Add<CustomInstallDialogs.ExitDialog>();

project.ManagedUI.ModifyDialogs.Add<CustomInstallDialogs.ModifyDialog>()
                                                       .Add<CustomInstallDialogs.FeaturesDialog>()
                                                       .Add<CustomInstallDialogs.ProgressDialog>()
                                                       .Add<CustomInstallDialogs.ExitDialog>();
```

 Above is my code for my CustomManagedUI.
It works fine while installing. But When run again after installation it shows up the modify dialog as expected, but the problem is that on clicking any of Repair/Remove/Change button the UI exits without any warning. 

> Feature dialog load started
> Feature dialog build feature tree started
> Feature dialog build feature tree completed
> Feature dialog load completed
> MSI (c) (80!BC) [14:28:25:395]: PROPERTY CHANGE: Adding WIXSHARP_RUNTIME_DATA property. Its value is 'WIXSHARP_MANAGED_UI_HANDLE=11409886'.
> Cannot attach ManagedUI: Microsoft.Deployment.WindowsInstaller.InstallCanceledException: Exception of type 'Microsoft.Deployment.WindowsInstaller.InstallCanceledException' was thrown.
>    at WixSharp.ManagedUI.Initialize(Session session, String resourcePath, InstallUIOptions& uiLevel)
> SFXCA: Initialization canceled by user.

above is the log section where an exception is seen. 

Complete log file is attached
[abc.txt](https://github.com/oleg-shilo/wixsharp/files/3706798/abc.txt)

**Comment by oleg-shilo** [comment]
You can try to debug your UI. 
. . .

When I tried the standard Custom UI template and it seems to work just fine.

I have attached the working sample. If it works in your environment, then the problem is in your project. Compare both and find the difference. This may help.

[WixSharp Setup3.zip](https://github.com/oleg-shilo/wixsharp/files/3712018/WixSharp.Setup3.zip)

**Comment by kriactivedesigns** [comment]
Actually I made my Custom UI by referring your sample project. The thing is that the UI works fine while installing, but it fails while running the msi again after installation, that is when the installer uses modify dialogs. The UI shows up but on pressing any action the UI goes to next screen and fails. I hope you could understand that from the log. 
.
Also I am not sure how to debug the UI while installer is executing. I am not that much familiar with windows forms but WPF. Can you please guide me how to debug the UI and also if possible please provide me a sample project for WPF XAML based custom UI

**Comment by oleg-shilo** [comment]
Before we proceed can you please run the sample that I sent you. Just to establish a benchmark: what works and what doesn't.

**Comment by kriactivedesigns** [comment]
Yes, I did run the sample code and it did work well. Also I used the Statandard managed  ui windows in my project and it also did work fine. The problem arises when I replace all managed ui windows with my custom managed uis and it fails only when the installer is run after installation, that is when the modify action is needed.

**Comment by oleg-shilo** [comment]
Then your environment is OK. Good starting point.

The problem is somehow is in your setup definition. The obvious suspect is your custom UI. That's why you need to debug it. And this is how you can do it.

1. Place `Debug.Assert(false);` in your dialog constructor. 
2. Run VS as admin. Elevation is required because MSI is always elevated so if one wants to debug it it also needs elevation.
3. Run your msi and observe assertion dialog.
4. Attach VS to the right instance of `rundll32.exe`. MSI never runs the custom actions themselves but rather executes them as a separate process `rundll32.exe`.

_You can try to do step #4 by simply selecting the VS instance from the assertion dialog but I found that sometimes it fails to attach properly. Though probably it makes sense to try it this way._

Now you are debugging. So see what throws the exception.

**Comment by kriactivedesigns** [comment]
[image: image.png]
[image: image.png]
This is what did I get. I am trying to figure this out, I expect your input
too

**Comment by kriactivedesigns** [comment]
Below is the assertion dialog shows up when i click change from the modify
dialog
[image: image.png]

**Comment by oleg-shilo** [comment]
Hm, I was really hoping for an exception :)

Well, let's see. Your environment is OK, Your dialog is also OK. This leaves only the deployment logic. 

I have checked your the only failure message "_Product: xxx -- Configuration failed._" and Google returned nothing real. 

I can only suggest almost a childish technique. Start switching off your deployment atrefacts one by one and see which one is causing the problem. 

Very long shot, any chance you are configuring some assemblies in GAC?

**Comment by kriactivedesigns** [comment]
```
private static void Main()
        {
            try
            {
                //This is to suppress the ICE57 warning related to the creation of Registry entry.
                Compiler.LightOptions += SetupConstants.Parameter_ICEWarning;

                List<string> ProductMSIs = new List<string>();
                ProductMSIs.Add(BuildDummyi4SuiteMSI());
                if (SetupConstants.IsBundlingNeeded)
                {
                    BuildBundle(ProductMSIs);
                }
            }
            catch (Exception ex)
            {
                Console.WriteLine("Build Failure");
                Console.WriteLine(ex.Message);
                Console.WriteLine(ex);
                throw new Exception("Build Failed", ex);
            }

        }

private static string BuildDummyi4SuiteMSI()
        {
            var project = new ManagedProject("DummyBuild",
                             new Dir(@"%ProgramFiles%\Dummy Company\DummyBuild",
                                 new File("Program.cs")));
            project.GUID = new Guid("6fe30b47-2577-43ad-9095-3861ba25889b");

            SetupCustomUI(project);
            return project.BuildMsi("DummyBuild/DummyBuild-x64");

        }

private static void SetupCustomUI(ManagedProject project)
        {
            project.LicenceFile = "Resources/i4Suite_License_Agreement.rtf";

            if (IsDefaultDilaogs)
            {
                project.ManagedUI = new ManagedUI();

                project.BannerImage = "Resources/topBannerBMP.bmp";
                project.BackgroundImage = "Resources/leftSideBannerBMP.bmp";

                project.ManagedUI.InstallDialogs.Add(Dialogs.Welcome)
                                                .Add(Dialogs.Licence)
                                                .Add(Dialogs.Features)
                                                .Add(Dialogs.Progress)
                                                .Add(Dialogs.Exit);

                project.ManagedUI.ModifyDialogs.Add(Dialogs.MaintenanceType)
                                                .Add(Dialogs.Features)
                                                .Add(Dialogs.Progress)
                                                .Add(Dialogs.Exit);

            }
            else
            {
                project.ManagedUI = new ManagedUI();

                project.ManagedUI.InstallDialogs.Add<CustomInstallDialogs.WelcomeDialog>()
                                                .Add<CustomInstallDialogs.LicenceDialog>()
                                                .Add<CustomInstallDialogs.FeaturesDialog>()
                                                .Add<CustomInstallDialogs.ProgressDialog>()
                                                .Add<CustomInstallDialogs.ExitDialog>();

                project.ManagedUI.ModifyDialogs.Add<CustomInstallDialogs.ModifyDialog>()
                                               .Add<CustomInstallDialogs.FeaturesDialog>()
                                               .Add<CustomInstallDialogs.ProgressDialog>()
                                               .Add<CustomInstallDialogs.ExitDialog>();

                //project.OutDir = "DummyBuild";
                //project.SourceBaseDir = ".";

                project.UILoaded += Project_UILoaded;
            }

        }

        private static void Project_UILoaded(SetupEventArgs e)
        {
            var managedUI = e.ManagedUI as Form;

            managedUI.AutoScaleMode = AutoScaleMode.Dpi;
            managedUI.PerformAutoScale();
            managedUI.FormBorderStyle = FormBorderStyle.FixedToolWindow;
            managedUI.StartPosition = FormStartPosition.CenterScreen;
            managedUI.Size = CustomManagedUI.Constants.WindowSize;
            e.ManagedUI.SetSize(CustomManagedUI.Constants.WindowSize.Width, CustomManagedUI.Constants.WindowSize.Height);
            managedUI.Refresh();
        }


```

This is how i am doing it in the Program.cs

**Comment by kriactivedesigns** [comment]
> using CustomManagedUI;
> using System;
> using System.Collections.Generic;
> using System.ComponentModel;
> using System.Data;
> using System.Diagnostics;
> using System.Drawing;
> using System.Linq;
> using System.Text;
> using System.Windows.Forms;
> using WixSharp;
> using WixSharp.UI.Forms;
> 
> namespace i4Suite.CustomInstallDialogs
> {
>     public partial class ModifyDialog : BaseDialogForm, IManagedDialog
>     {
>         public ModifyDialog()
>         {
>             Debug.Assert(false);
>             InitializeComponent();
>         }
> 
>         Type ProgressDialog
>         {
>             get
>             {
>                 return Shell.Dialogs
>                             .Where(d => d.GetInterfaces().Contains(typeof(IProgressDialog)))
>                             .FirstOrDefault();
>             }
>         }
> 
>         private void ModifyDialog_Load(object sender, EventArgs e)
>         {
>             modifyBanner.Title = "Modifying";
>             modifyBanner.Description = Constants.ModifyDialogDescription;
> 
>             #region Create and Assign ClickHandler To Buttons
>             Button repairButton = new Button
>             {
>                 Text = "Repair",
>                 TextAlign = System.Drawing.ContentAlignment.MiddleCenter
>             };
>             Button changeButton = new Button
>             {
>                 Text = "Change",
>                 TextAlign = System.Drawing.ContentAlignment.MiddleCenter
>             };
>             Button removeButton = new Button
>             {
>                 Text = "Remove",
>                 TextAlign = System.Drawing.ContentAlignment.MiddleCenter
>             };
> 
>             repairButton.Click += repair_Click;
>             changeButton.Click += change_Click;
>             removeButton.Click += remove_Click;
>             #endregion
> 
>             #region Change
>             tableLayoutPanel.Controls.Add(new Label()
>             {
>                 Text = Constants.ChangeDescription,
>                 TextAlign = System.Drawing.ContentAlignment.MiddleRight
>             }, 0, 0);
> 
>             tableLayoutPanel.Controls.Add( changeButton, 0,1);
>             #endregion
> 
>             #region Repair
>             tableLayoutPanel.Controls.Add(new Label()
>             {
>                 Text = Constants.RepairDescription,
>                 TextAlign = System.Drawing.ContentAlignment.MiddleRight
>             }, 0, 0);
> 
>             tableLayoutPanel.Controls.Add( repairButton, 1, 1);
>             #endregion
> 
>             #region Remove
>             tableLayoutPanel.Controls.Add(new Label()
>             {
>                 Text = Constants.removeDescription,
>                 TextAlign = System.Drawing.ContentAlignment.MiddleRight
>             }, 0, 0);
> 
>             tableLayoutPanel.Controls.Add( removeButton, 2, 1);
>             #endregion
>         }
> 
>         void change_Click(object sender, System.EventArgs e)
>         {
>             Runtime.Session["MODIFY_ACTION"] = "Change";
>             Shell.GoNext();
>         }
> 
>         void repair_Click(object sender, System.EventArgs e)
>         {
>             Runtime.Session["MODIFY_ACTION"] = "Repair";
>             int index = Shell.Dialogs.IndexOf(ProgressDialog);
>             if (index != -1)
>                 Shell.GoTo(index);
>             else
>                 Shell.GoNext();
>         }
> 
>         void remove_Click(object sender, System.EventArgs e)
>         {
>             Runtime.Session["REMOVE"] = "ALL";
>             Runtime.Session["MODIFY_ACTION"] = "Remove";
> 
>             int index = Shell.Dialogs.IndexOf(ProgressDialog);
>             if (index != -1)
>                 Shell.GoTo(index);
>             else
>                 Shell.GoNext();
>         }
> 
>         private void cancelButton_Click(object sender, EventArgs e)
>         {
>             DialogResult messageResult = MessageBox.Show(this, Constants.WarningPopUpMessage, Constants.WarningPopUpCaption, MessageBoxButtons.YesNo);
>             if (messageResult == DialogResult.Yes)
>             {
>                 Shell.Cancel();
>             }
>             else
>                 return;
>         }
>     }
> }
> 

This is my modify dialog

**Comment by kriactivedesigns** [comment]
> Cannot attach ManagedUI: Microsoft.Deployment.WindowsInstaller.InstallCanceledException: Exception of type 'Microsoft.Deployment.WindowsInstaller.InstallCanceledException' was thrown.
> at WixSharp.ManagedUI.Initialize(Session session, String resourcePath, InstallUIOptions& uiLevel)
> SFXCA: Initialization canceled by user.

.
as of my understanding ( may be I am wrong) InstallCanceledException is thrown from the below code in the https://github.com/oleg-shilo/wixsharp/blob/master/Source/src/WixSharp.UI/ManagedUI/ManagedUI.cs 
```
if (waitResult == 1)
                    {
                        //UI exited without starting the install. Cancel the installation.
                        throw new InstallCanceledException();
                    }
```

**Comment by kriactivedesigns** [comment]
//  Start switching off your deployment atrefacts // sorry that i didnt
understand what you really meant by that.

**Comment by oleg-shilo** [comment]
What I meant is that. You have a benchmark project "A" that works well. You also have another project "B" that it was derived from project "A" but has some additional features/routines/classes. These additions make this project fail. Start excluding them one by one until and see at what point your project starts working. You may end up pretty much with project "B" becoming an equivalent of project "A". But that will give you the answer you are looking for anyway.

The exception you quoted indicates that the user exited the UI sequence without proceeding to the next step. And indeed you do not have any on_click handler for the "Next" button.

After looking closer at your custom dialog.... I am not sure how you ended up with it. It is not consistent with WixSharp samples, it is not consistent with the "CustomDialog" VS project template.

I suggest you scrap it and create a new project from the recommended VS project template "WixSharp Managed Project - Custom UI". Verify that it works as expected and then proceed with adding your features and verifying the health of the project after adding every significant modification.

Alternatively you can try to revive your dialog code (that you posted here). Just reintroduce "Next" button. Though I am reluctant to advice you going this way.

**Comment by kriactivedesigns** [comment]
I have done more debugging. What I found is that The issue happens only
when I go to the next screen ( whichever ) from the ModifyDialog it fails.
Even I just use Shell.GoNext().  I tried putting a dummy dialog before and
after the ModifyDialog and I could navigate from first screen to the modify
dialog but when tried to navigate from the ModifyDialog to the next one it
failed. Also I tried using the Default FeatureDialog inplace of my Custom
Dialog and put my Custom Modify dialog before it and custom progress dialog
after it. then I could go to feature dialog without any problem but when I
clicked the next button it failed.

below is the log of that test

SFXCA: Extracting embedded UI to temporary directory:
C:\Users\arun\AppData\Local\Temp\MSI27436
SFXCA: Binding to CLR version v4.0.30319
Cannot attach ManagedUI: System.ObjectDisposedException: Cannot access a
disposed object.
Object name: 'ShellView'.
   at System.Windows.Forms.Control.MarshaledInvoke(Control caller, Delegate
method, Object args, Boolean synchronous)
   at System.Windows.Forms.Control.Invoke(Delegate method, Object args)
   at System.Windows.Forms.Control.Invoke(Delegate method)
   at WixSharp.UIShell.InUIThread(Action action)
   at WixSharp.UIShell.OnExecuteStarted()
   at WixSharp.ManagedUI.Initialize(Session session, String resourcePath,
InstallUIOptions& uiLevel)
Exception thrown by embedded UI initialization:
System.ObjectDisposedException: Cannot access a disposed object.
Object name: 'ShellView'.
   at System.Windows.Forms.Control.MarshaledInvoke(Control caller, Delegate
method, Object args, Boolean synchronous)
   at System.Windows.Forms.Control.Invoke(Delegate method, Object args)
   at System.Windows.Forms.Control.Invoke(Delegate method)
   at WixSharp.UIShell.InUIThread(Action action)
   at WixSharp.UIShell.OnExecuteStarted()
   at WixSharp.ManagedUI.Initialize(Session session, String resourcePath,
InstallUIOptions& uiLevel)
   at
Microsoft.Deployment.WindowsInstaller.EmbeddedUIProxy.Initialize(Int32
sessionHandle, String uiClass, Int32 internalUILevel)
Info 1603. Fatal error during installation.
D:\Spia\SpiaActive\Setup\WixSharp\i4Suite\i4Suite\DummyBuild\DummyBuild-x64.msi
MSI (c) (68:FC) [14:31:31:322]: Product: DummyBuild -- Configuration failed.

MSI (c) (68:FC) [14:31:31:323]: Windows Installer reconfigured the product.
Product Name: DummyBuild. Product Version: 1.0.0.0. Product Language: 1033.
Manufacturer: ARUN. Reconfiguration success or error status: 1603.
